crates/adapters/src/controller/error.rs

```diff
@@ -35,38 +44,68 @@ pub enum ConfigError {
     UnknownOutputStream { stream_name: String },
 }
 
+impl StdError for ConfigError {}
+
+impl DetailedError for ConfigError {
+    fn error_code(&self) -> Cow<'static, str> {
+        match self {
+            Self::ConfigParseError { .. } => Cow::from("ConfigParseError"),
+            Self::DuplicateInputEndpoint { .. } => Cow::from("DuplicateInputEndpoint"),
+            Self::DuplicateOutputEndpoint { .. } => Cow::from("DuplicateOutputEndpoint"),
+            Self::UnknownInputFormat { .. } => Cow::from("UnknownInputFormat"),
+            Self::UnknownOutputFormat { .. } => Cow::from("UnknownOutputFormat"),
+            Self::UnknownInputTransport { .. } => Cow::from("UnknownInputTransport"),
+            Self::UnknownOutputTransport { .. } => Cow::from("UnknownOutputTransport"),
+            Self::UnknownInputStream { .. } => Cow::from("UnknownInputStream"),
+            Self::UnknownOutputStream { .. } => Cow::from("UnknownOutputStream"),
+        }
+    }
+}
+
 impl Display for ConfigError {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
         match self {
+            Self::ConfigParseError { error } => {
```

Comment:
My suggestion would be to use the thiserror macro https://docs.rs/thiserror/latest/thiserror/ to generate this Display stuff with annotations

Reply:
Another clever crate I should use :man_facepalming: